pkg/scheduler/strategy/rest_by_job.go

```diff
@@ -0,0 +1,128 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategy
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	prowv1 "sigs.k8s.io/prow/pkg/apis/prowjobs/v1"
+	"sigs.k8s.io/prow/pkg/config"
+)
+
+// SchedulingRequest represents the incoming request structure
+type SchedulingRequest struct {
+	Job string `json:"job"`
+}
+
+// SchedulingResponse represents the response structure
+type SchedulingResponse struct {
+	Cluster string `json:"cluster"`
+}
+
+type cacheEntry struct {
+	r         SchedulingResponse
+	timestamp time.Time
+}
+
+// RestByJob is a strategy that schedules a ProwJob on a specific cluster
+// based on the response of the ProwJob's Job name to given URL
+type RESTByJob struct {
+	cfg       config.RESTByJobScheduling
+	cache     map[string]*cacheEntry
+	timestamp time.Time
+}
+
+// NewRESTByJob creates a new RestByJob instance with caching
+func NewRESTByJob(cfg config.RESTByJobScheduling) *RESTByJob {
+	rbj := &RESTByJob{
+		cfg:   cfg,
+		cache: make(map[string]*cacheEntry),
+	}
+	return rbj
+}
+
+// query sends a POST request to the specified URL with the provided request body
+// and returns the response as a SchedulingResponse object
+func query(url string, reqBody SchedulingRequest) (SchedulingResponse, error) {
+	jsonData, err := json.Marshal(reqBody)
+	if err != nil {
+		return SchedulingResponse{}, fmt.Errorf("error marshaling JSON: %w", err)
+	}
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return SchedulingResponse{}, fmt.Errorf("error creating request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return SchedulingResponse{}, fmt.Errorf("error sending request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return SchedulingResponse{}, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	var respBody SchedulingResponse
+	if err := json.NewDecoder(resp.Body).Decode(&respBody); err != nil {
+		return SchedulingResponse{}, fmt.Errorf("error decoding response: %w", err)
+	}
+
+	return respBody, nil
+}
+
+// Schedule checks the cache for a valid response or queries the REST API if the cache is stale
+func (rbj *RESTByJob) Schedule(ctx context.Context, pj *prowv1.ProwJob) (Result, error) {
+	rbj.cleanupCache()
+	entry, found := rbj.cache[pj.Spec.Job]
+
+	if found && time.Since(entry.timestamp) < rbj.cfg.Timeout.Duration {
+		return Result(entry.r), nil
+	}
+
+	resp, err := query(rbj.cfg.URL, SchedulingRequest{Job: pj.Spec.Job})
+	if err != nil {
+		return Result{}, fmt.Errorf("error querying URL: %w", err)
+	}
+
+	rbj.cache[pj.Spec.Job] = &cacheEntry{
+		r:         resp,
+		timestamp: time.Now(),
+	}
+	return Result(resp), nil
+}
+
+// cleanupCache removes stale entries from the cache
+func (rbj *RESTByJob) cleanupCache() {
```

Comment:
This is running when a job is scheduled. Why not have another mechanism where the cache is being cleaned after every N time?

Reply:
It's invoking this function every time, yes, but if you look at the implementation of the fcn, in most of the cases function quits immediately as interval is not met. So, effectively, it will clean map every `n` hour. I think this is better then after every n time, more understandable for the user.

Reply:
I'll change it to be more visible (move to Schedule fcn).